### ISSUE TITLE
undefined symbol error when importing velocyto

### ISSUE THREAD

**Issue by samfux84** [velocyto-team/velocyto.py#186]
Hi,

I have installed velocyto with pip for Python 3.7.1, but when importing the module, it fails with an error message about an undefined symbol:

```
[sfux@eu-login-08-ng ~]$ python
Python 3.7.1 (default, Mar 14 2019, 07:27:59) 
[GCC 4.8.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import velocyto
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/velocyto/__init__.py", line 13, in <module>
    from .estimation import fit_slope, _fit1_slope, clusters_stats
  File "/cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/velocyto/estimation.py", line 7, in <module>
    from .speedboosted import _colDeltaCor, _colDeltaCorLog10, _colDeltaCorSqrt
ImportError: /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/velocyto/speedboosted.cpython-37m-x86_64-linux-gnu.so: undefined symbol: __kmpc_dispatch_next_4
>>> quit()
[sfux@eu-login-08-ng ~]$
```

Is this a known problem ?

**Comment by gioelelm** [comment]
It is likely that there was a problem at the C (or cython) compilation. I would suggest to unistall, make sure cython, gcc, libgcc, llvm, gomp are installed (using conda) correctly and then reinstall. Sent from my mobile - excuse the typos

**Comment by samfux84** [comment]
@gioelelm : Thank you for your reply. On our HPC cluster, we do not support conda.

I have tried uninstalling the package and rebuilding it (not using the precompiled wheel), but it did not solve the problem:

```
[sfux@eu-login-06-ng ~]$ pip uninstall velocyto
Uninstalling velocyto-0.17.17:
  Would remove:
    /cluster/apps/python/3.7.1/x86_64/bin/velocyto
    /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/velocyto-0.17.17-py3.7.egg-info
    /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/velocyto/*
Proceed (y/n)? y
  Successfully uninstalled velocyto-0.17.17
[sfux@eu-login-06-ng ~]$ pip install --no-binary :all: velocyto
Collecting velocyto
  Using cached https://files.pythonhosted.org/packages/81/66/e8fff9d3b824fd99c0f678c47c740fec058ce2d9a0cfdf11b114ea8889f2/velocyto-0.17.17.tar.gz
Requirement already satisfied: numpy in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (1.16.2)
Requirement already satisfied: scipy in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (1.2.1)
Requirement already satisfied: cython in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (0.29.6)
Requirement already satisfied: numba in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (0.43.0)
Requirement already satisfied: matplotlib in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (3.0.3)
Requirement already satisfied: scikit-learn in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (0.20.3)
Requirement already satisfied: h5py in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (2.9.0)
Requirement already satisfied: loompy in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (2.0.17)
Requirement already satisfied: pysam in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (0.15.2)
Requirement already satisfied: Click in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (7.0)
Requirement already satisfied: pandas in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto) (0.24.1)
Requirement already satisfied: llvmlite>=0.28.0dev0 in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from numba->velocyto) (0.28.0)
Requirement already satisfied: cycler>=0.10 in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from matplotlib->velocyto) (0.10.0)
Requirement already satisfied: kiwisolver>=1.0.1 in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from matplotlib->velocyto) (1.0.1)
Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from matplotlib->velocyto) (2.3.1)
Requirement already satisfied: python-dateutil>=2.1 in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from matplotlib->velocyto) (2.8.0)
Requirement already satisfied: six in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from h5py->velocyto) (1.12.0)
Requirement already satisfied: setuptools in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from loompy->velocyto) (40.8.0)
Requirement already satisfied: pytz>=2011k in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from pandas->velocyto) (2018.9)
Skipping bdist_wheel for velocyto, due to binaries being disabled for it.
Installing collected packages: velocyto
  Running setup.py install for velocyto ... done
Successfully installed velocyto-0.17.17
[sfux@eu-login-06-ng ~]$ python
Python 3.7.1 (default, Mar 14 2019, 07:27:59) 
[GCC 4.8.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import velocyto
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/velocyto/__init__.py", line 13, in <module>
    from .estimation import fit_slope, _fit1_slope, clusters_stats
  File "/cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/velocyto/estimation.py", line 7, in <module>
    from .speedboosted import _colDeltaCor, _colDeltaCorLog10, _colDeltaCorSqrt
ImportError: /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/velocyto/speedboosted.cpython-37m-x86_64-linux-gnu.so: undefined symbol: __kmpc_dispatch_next_4
>>>
```
To build Velocyto, I am using:

- GCC 4.8.2
- LLVM 7.0.1
- numpy 1.16.2
- scipy 1.2.1
- cython  0.29.6
- numba 0.43.0
- matplotlib 3.0.3
- scikit-learn 0.20.3
- h5py 2.9.0
- click 7.0
- pysam 0.15.2

I installed all the dependencies myself using pip (with the option "--no-binary :all:" to make sure everything is built).

A google search of the error message did not really help.

Any help is appreciated.

**Comment by gioelelm** [comment]
The problem seems to be a failure at importing a compiled extension that is part of velocyto (speedbusted) that is written in Cython and compiled by the C compiler using OpenMP parallelisation.
__kmp_dispatch_next_4 appears to be some kind of routine of OpenMP, therefore the error is definitely pointing in that direction.
I think you might want to make sure that the gcc compiler version supports Open MP or that pip is linked to the correct version of the compiler when installing.

**Comment by samfux84** [comment]
I also tried to build from the most recent source:

```
[sfux@eu-login-06-ng velocyto.py-0.17.17]$ pip install -v -e .   
Created temporary directory: /tmp/pip-ephem-wheel-cache-c1fsfkzf
Created temporary directory: /tmp/pip-req-tracker-190nl1xq
Created requirements tracker '/tmp/pip-req-tracker-190nl1xq'
Created temporary directory: /tmp/pip-install-73du_s_f
Obtaining file:///cluster/home/sfux/test/velocyto/velocyto.py-0.17.17
  Added file:///cluster/home/sfux/test/velocyto/velocyto.py-0.17.17 to build tracker '/tmp/pip-req-tracker-190nl1xq'
    Running setup.py (path:/cluster/home/sfux/test/velocyto/velocyto.py-0.17.17/setup.py) egg_info for package from file:///cluster/home/sfux/test/velocyto/velocyto.py-0.17.17
    Running command python setup.py egg_info
    ['-c', 'egg_info']
    running egg_info
    writing velocyto.egg-info/PKG-INFO
    writing dependency_links to velocyto.egg-info/dependency_links.txt
    writing entry points to velocyto.egg-info/entry_points.txt
    writing requirements to velocyto.egg-info/requires.txt
    writing top-level names to velocyto.egg-info/top_level.txt
    reading manifest file 'velocyto.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    writing manifest file 'velocyto.egg-info/SOURCES.txt'
  Source in /cluster/home/sfux/test/velocyto/velocyto.py-0.17.17 has version 0.17.17, which satisfies requirement velocyto==0.17.17 from file:///cluster/home/sfux/test/velocyto/velocyto.py-0.17.17
  Removed velocyto==0.17.17 from file:///cluster/home/sfux/test/velocyto/velocyto.py-0.17.17 from build tracker '/tmp/pip-req-tracker-190nl1xq'
Requirement already satisfied: numpy in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (1.16.2)
Requirement already satisfied: scipy in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (1.2.1)
Requirement already satisfied: cython in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (0.29.6)
Requirement already satisfied: numba in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (0.43.0)
Requirement already satisfied: matplotlib in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (3.0.3)
Requirement already satisfied: scikit-learn in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (0.20.3)
Requirement already satisfied: h5py in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (2.9.0)
Requirement already satisfied: loompy in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (2.0.17)
Requirement already satisfied: pysam in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (0.15.2)
Requirement already satisfied: Click in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (7.0)
Requirement already satisfied: pandas in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from velocyto==0.17.17) (0.24.1)
Requirement already satisfied: llvmlite>=0.28.0dev0 in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from numba->velocyto==0.17.17) (0.28.0)
Requirement already satisfied: cycler>=0.10 in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from matplotlib->velocyto==0.17.17) (0.10.0)
Requirement already satisfied: kiwisolver>=1.0.1 in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from matplotlib->velocyto==0.17.17) (1.0.1)
Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from matplotlib->velocyto==0.17.17) (2.3.1)
Requirement already satisfied: python-dateutil>=2.1 in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from matplotlib->velocyto==0.17.17) (2.8.0)
Requirement already satisfied: six in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from h5py->velocyto==0.17.17) (1.12.0)
Requirement already satisfied: setuptools in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from loompy->velocyto==0.17.17) (40.8.0)
Requirement already satisfied: pytz>=2011k in /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages (from pandas->velocyto==0.17.17) (2018.9)
Installing collected packages: velocyto
  Running setup.py develop for velocyto
    Running command /cluster/apps/python/3.7.1/x86_64/bin/python3.7 -c "import setuptools, tokenize;__file__='/cluster/home/sfux/test/velocyto/velocyto.py-0.17.17/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" develop --no-deps
    ['-c', 'develop', '--no-deps']
    running develop
    running egg_info
    writing velocyto.egg-info/PKG-INFO
    writing dependency_links to velocyto.egg-info/dependency_links.txt
    writing entry points to velocyto.egg-info/entry_points.txt
    writing requirements to velocyto.egg-info/requires.txt
    writing top-level names to velocyto.egg-info/top_level.txt
    reading manifest file 'velocyto.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    writing manifest file 'velocyto.egg-info/SOURCES.txt'
    running build_ext
    building 'velocyto.speedboosted' extension
    creating build
    creating build/temp.linux-x86_64-3.7
    creating build/temp.linux-x86_64-3.7/velocyto
    /cluster/apps/llvm/7.0.1/x86_64/bin/clang -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -ftree-vectorize -march=core-avx2 -mavx2 -ftree-vectorize -march=core-avx2 -mavx2 -fPIC -I/cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/numpy/core/include -I/cluster/apps/python/3.7.1/x86_64/include/python3.7m -I/cluster/apps/python/3.7.1/x86_64/include/python3.7m -c velocyto/speedboosted.c -o build/temp.linux-x86_64-3.7/velocyto/speedboosted.o -fopenmp -ffast-math
    In file included from velocyto/speedboosted.c:545:
    In file included from /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4:
    In file included from /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12:
    In file included from /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1822:
    /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: "Using deprecated NumPy API, disable it with "          "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-W#warnings]
    #warning "Using deprecated NumPy API, disable it with " \
     ^
    1 warning generated.
    creating build/lib.linux-x86_64-3.7
    creating build/lib.linux-x86_64-3.7/velocyto
    /cluster/apps/gcc/4.8.2/bin/gcc -pthread -shared build/temp.linux-x86_64-3.7/velocyto/speedboosted.o -L/cluster/apps/python/3.7.1/x86_64/lib64 -lpython3.7m -o build/lib.linux-x86_64-3.7/velocyto/speedboosted.cpython-37m-x86_64-linux-gnu.so -fopenmp
    copying build/lib.linux-x86_64-3.7/velocyto/speedboosted.cpython-37m-x86_64-linux-gnu.so -> velocyto
    Creating /cluster/apps/python/3.7.1/x86_64/lib64/python3.7/site-packages/velocyto.egg-link (link to .)
    Adding velocyto 0.17.17 to easy-install.pth file
    Installing velocyto script to /cluster/apps/python/3.7.1/x86_64/bin

    Installed /cluster/home/sfux/test/velocyto/velocyto.py-0.17.17
Successfully installed velocyto
Cleaning up...
Removed build tracker '/tmp/pip-req-tracker-190nl1xq'
[sfux@eu-login-06-ng velocyto.py-0.17.17]$ python
Python 3.7.1 (default, Mar 14 2019, 07:27:59) 
[GCC 4.8.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import velocyto
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/cluster/home/sfux/test/velocyto/velocyto.py-0.17.17/velocyto/__init__.py", line 13, in <module>
    from .estimation import fit_slope, _fit1_slope, clusters_stats
  File "/cluster/home/sfux/test/velocyto/velocyto.py-0.17.17/velocyto/estimation.py", line 7, in <module>
    from .speedboosted import _colDeltaCor, _colDeltaCorLog10, _colDeltaCorSqrt
ImportError: /cluster/home/sfux/test/velocyto/velocyto.py-0.17.17/velocyto/speedboosted.cpython-37m-x86_64-linux-gnu.so: undefined symbol: __kmpc_dispatch_next_4
>>>
```
To me, it looks like speedboosted.cpython-37m-x86_64-linux-gnu.so was linked against libgomp from LLVM 7.0.1, but at runtime, it is using gomp from GCC.

libgomp from LLVM 7.0.1 contains the missing symbols:

```
[sfux@eu-login-21-ng velocyto]$ readelf -a /cluster/apps/llvm/7.0.1/x86_64/lib/libgomp.so | grep __kmpc_dispatch_next_4
   200: 0000000000086e60  1027 FUNC    GLOBAL DEFAULT   12 __kmpc_dispatch_next_4@@VERSION
   631: 0000000000087bb0  1027 FUNC    GLOBAL DEFAULT   12 __kmpc_dispatch_next_4u@@VERSION
  2166: 0000000000086e60  1027 FUNC    GLOBAL DEFAULT   12 __kmpc_dispatch_next_4
  2596: 0000000000087bb0  1027 FUNC    GLOBAL DEFAULT   12 __kmpc_dispatch_next_4u
```

**Comment by ggruenhagen3** [comment]
I had this same error and spent hours trying to fix it using solutions posted here and elsewhere mostly involving conda compilers. In the end, I was able to get velocyto to work by installing from source (it actually turned out to be an easy and simple solution).
